runtime/compiler/optimizer/JProfilingBlock.cpp

```diff
@@ -902,37 +887,59 @@ void TR_JProfilingBlock::addRecompilationTests(TR_BlockFrequencyInfo *blockFrequ
                }
             }
          }
-      TR::Node *root = addRoot;
+      root = addRoot;
       if (subRoot)
          {
          root = TR::Node::create(node, TR::isub, 2, root, subRoot);
          }
+      }
+   return root;
+   }
+
+/**
+ * Add runtime tests to the start of the method to trigger method reccompilation once the
+ * appropriate number of method entries has occurred as determined by the raw block
+ * count of the first block of the method.
+ */   
+void TR_JProfilingBlock::addRecompilationTests(TR_BlockFrequencyInfo *blockFrequencyInfo, TR_BitVector **componentCounters)
+   {
+   int32_t startBlockNumber = comp()->getStartBlock()->getNumber();
+   blockFrequencyInfo->setEntryBlockNumber(startBlockNumber);
 
+   TR::Node *node = comp()->getMethodSymbol()->getFirstTreeTop()->getNode();
+   TR::Node *root = generateBlockRawCountCalculationNode(comp(), startBlockNumber, node, blockFrequencyInfo, componentCounters);
+   if (root != NULL)
+      {
       TR::Block * originalFirstBlock = comp()->getStartBlock();
 
       TR::Block *guardBlock1 = TR::Block::createEmptyBlock(node, comp(), originalFirstBlock->getFrequency());
          {
-         TR::SymbolReference *symRef = comp()->getSymRefTab()->createKnownStaticDataSymbolRef(blockFrequencyInfo->getEnableJProfilingRecompilation(), TR::Int32);
-         TR::Node *enableLoad = TR::Node::createWithSymRef(node, TR::iload, 0, symRef);
-         TR::Node *enableTest = TR::Node::createif(TR::ificmpeq, enableLoad, TR::Node::iconst(node, 0), originalFirstBlock->getEntry());
+         TR::Node *enableLoad = TR::Node::createWithSymRef(node, TR::iload, 0, blockFrequencyInfo->getOrCreateSymRefForIsQueuedForRecompilation(comp()));
+         TR::Node *enableTest = TR::Node::createif(TR::ificmpeq, enableLoad, TR::Node::iconst(node, -1), originalFirstBlock->getEntry());
          TR::TreeTop *enableTree = TR::TreeTop::create(comp(), enableTest);
          enableTest->setIsProfilingCode();
          guardBlock1->append(enableTree);
          }
 
+      static int32_t profilingCompileThreshold = comp()->getOptions()->getJProfilingMethodRecompThreshold();
+
+      traceMsg(comp(),"Profiling Compile Threshold for method = %d\n", profilingCompileThreshold);
       TR::Block *guardBlock2 = TR::Block::createEmptyBlock(node, comp(), originalFirstBlock->getFrequency());
-      TR::Node *recompThreshold = TR::Node::createWithSymRef(node, TR::iload, 0, comp()->getSymRefTab()->createKnownStaticDataSymbolRef(thresholdLocation, TR::Int32));
+      TR::Node *recompThreshold = TR::Node::iconst(node, profilingCompileThreshold);
```

Comment:
Why has this been made a constant unconditionally - being able to vary the compile threshold is certainly something we would want to consider - for a profiled compile the answer may be static but for a non-profiled compile we may need some ability to vary things so I'm not sure that this is an 'improvement' - seems a bit of a hack.

Reply:
Changed in https://github.com/eclipse/openj9/pull/2348/commits/849a86a8dee5b163a674b092070399883c69e384